doc/design.md

```diff
@@ -0,0 +1,949 @@
+A considerable number of algorithms are checking this manually, branching to different code sections.
+`HistogramData::Histogram` attempted to solve this by introducing getters like `Histogram::counts()` and `Histogram::frequencies()`.
+However, it seems to be useful only in a rather limited number of places, whereas it adds a significant amount of complication to the API.
+The flag adds complication and should thus be removed.
+
+##### Implementation
+
+This will require having a unit for "counts", i.e., we cannot simply use "dimensionless".
+We need to establish whether it is possible to write clean client code based on the unit-driven mechanism.
+Apart from that there should not be any major implementation difficulties.
+
+
+### <a name="design-details-no-progress-reporting-no-cancellation"></a>No progress reporting and no cancellation
+
+##### Rationale
+
+- Most operations are simple and thus fast.
+- Cancellation and progress reporting can be implemented in higher-level algorithms that *use* `Dataset`.
+- Cancellation leads to undefined side effects for in-place operation, i.e., it breaks data.
+- Adds complexity and complicates the interface.
+- `numpy` operations cannot be cancelled and do not report progress.
+
+
+### <a name="design-details-no-length-variation"></a>No support for histograms with varying length
+
+##### Rationale
+
+This has not been supported in `DataObjects::Workspace2D` for 10 years, until a very recent change, i.e., it does not appear to be strictly necessary.
+Support has been considered and prototyped, however there are two strong reasons it is not part of the final design:
+- Significant complication of implementation for all operations, leading to larger initial development cost and more difficult long-term maintenance.
+- Negatively affects the performance of iteration via `DatasetView`, *even if the length does not vary*.
+
+##### Workaround
+
+It is possible to store a variable holding datasets in a dataset.
+The result could be very similar to a current `DataObjects::Workspace2D`.
+In general we should discourage doing this since most of the other functionality we provide would not work nicely with such as nested dataset.
+However, if a varying histogram length is essential for some purpose it is absolutely possible.
+
+
+### <a name="design-details-threading"></a>Operations are threaded using OpenMP
+
+##### Rationale
+
+Threading using OpenMP in a fork-join style manner is not the most efficient one.
+As part of the prototyping process we considered other solutions, in particular [dask](https://dask.org/).
+Dask could solve many problems such as overcoming the limitations of the fork-join approach as providing a unified solution for threading and multi-process parallelisation such as MPI.
+However, the prototyping experience indicated that it is not performant enough to completely replace threading for our typical workloads.
```

Comment:
Based on time scales to implement this, should we monitor if progress eliminating overheads in dask is made?

Reply:
I think `dask` is not off the table for replacing what we are doing, e.g., with MPI, so yes we will keep looking at it. However, since it is not a solution for threading this would be outside the scope of this design.